File: rust/shm-bridge/src/seqlock.rs
//! The seqlock protocol from `backend/internal/shm/{writer,reader}.go`,
//! generalised over the two segment types.
//!
//! Wire protocol (unchanged, shared with Go and the IEC runtime):
//! the writer makes `Header.Seq` odd, mutates the payload, then makes it even
//! again; a reader that sees an odd seq, or different seqs before and after
//! its copy, retries. The daemon is the *writer* of `plc_data` and the
//! *reader* of `plc_cmd`.
//!
//! Memory model: unlike Go (whose bulk struct copy is racy under C11 rules
//! but tolerated by its runtime and x86-TSO), this implementation is fully
//! defined under the Rust/C11 model: the payload is copied as relaxed
//! per-`u64` atomic accesses and the seq transitions carry acquire/release
//! fences (the fence-based seqlock from Boehm, "Can seqlocks get along with
//! programming language memory models?"). On x86-64 all of it compiles to
//! plain MOVs plus compiler barriers — zero runtime cost, same bytes on the
//! wire.

use core::mem::size_of;
use core::sync::atomic::{fence, AtomicU32, AtomicU64, Ordering};

use crate::layout::{Header, Segment};
use crate::mapping::Mapping;

/// Mirror of Go's `seqlockMaxRetries` and the IEC reader's `FOR i := 0 TO 99`.
pub const SEQLOCK_MAX_RETRIES: usize = 100;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum ReadError {
    /// Writer kept us spinning (`ErrSeqlockBusy`).
    Busy,
    /// Wrong segment or layout (`ErrMagicMismatch`).
    MagicMismatch,
    /// Rebuild sides with matching layout (`ErrVersionMismatch`).
    VersionMismatch,
}

impl core::fmt::Display for ReadError {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::Busy => write!(f, "seqlock: writer kept us spinning"),
            Self::MagicMismatch => write!(f, "magic mismatch: wrong segment or layout"),
            Self::VersionMismatch => {
                write!(f, "version mismatch: rebuild peers with matching layout")
            }
        }
    }
}

impl std::error::Error for ReadError {}

const SEQ_OFFSET: usize = core::mem::offset_of!(Header, seq);
/// Index of the u64 word that contains `Header.Seq` (+ its padding). The
/// bulk copy loops skip it: the seq word is owned by the AtomicU32 below,
/// and mixing atomic access sizes on the same location is not defined.
const SEQ_WORD: usize = SEQ_OFFSET / 8;

fn seq_atomic(m: &Mapping) -> &AtomicU32 {
    debug_assert!(m.len() >= size_of::<Header>());
    // Safety: offset 8 is in bounds, 4-aligned (mapping is 8-aligned), and
    // every access to this location goes through atomics.
    unsafe { AtomicU32::from_ptr(m.ptr().add(SEQ_OFFSET) as *mut u32) }
}

/// Publish `src` to the segment under the seqlock protocol — the daemon-side
/// counterpart of Go's `WritePlcCommand` (same protocol, opposite segment).
///
/// The writer — not the caller — owns `Header.Seq`; whatever seq the caller
/// left in `src` is ignored, and the payload copy never carries an even seq
/// while the payload is in flux (see writer.go's comment for the tear this
/// prevents).
pub fn publish<T: Segment>(m: &Mapping, src: &T) {
    assert!(m.len() >= size_of::<T>(), "segment smaller than {}", T::NAME);
    let seq = seq_atomic(m);

    let s = seq.load(Ordering::Relaxed);
    let odd = s.wrapping_add(1); // sole writer: s is even, s+1 is odd
    seq.store(odd, Ordering::Relaxed);
    fence(Ordering::Release); // begin-write: payload stores stay after the odd store

    let mut tmp = *src;
    tmp.header_mut().seq = odd;
    tmp.header_mut()._pad = 0;
    let words = size_of::<T>() / 8;
    let src_words = &tmp as *const T as *const u64;
    let dst = m.ptr() as *mut u64;
    for i in 0..words {
        if i == SEQ_WORD {
            continue; // seq word is owned by the AtomicU32
        }
        // Safety: in bounds (asserted above), 8-aligned, atomic.
        unsafe { AtomicU64::from_ptr(dst.add(i)).store(src_words.add(i).read(), Ordering::Relaxed) };
    }

    seq.store(odd.wrapping_add(1), Ordering::Release); // end-write: even == stable
}

/// Copy the segment into `dst` under the seqlock protocol — the daemon-side
/// counterpart of Go's `ReadPlcData` (same protocol, opposite segment).
///
/// On success `dst.header.magic/version` are guaranteed to match this
/// build's expectations, exactly like the Go reader.
pub fn snapshot<T: Segment>(m: &Mapping, dst: &mut T) -> Result<(), ReadError> {
    assert!(m.len() >= size_of::<T>(), "segment smaller than {}", T::NAME);
    let seq = seq_atomic(m);
    let words = size_of::<T>() / 8;

    for _ in 0..SEQLOCK_MAX_RETRIES {
        let s1 = seq.load(Ordering::Acquire);
        if s1 & 1 != 0 {
            continue;
        }
        let src = m.ptr() as *const u64;
        let dst_words = dst as *mut T as *mut u64;
        for i in 0..words {
            if i == SEQ_WORD {
                continue;
            }
            // Safety: in bounds (asserted above), 8-aligned, atomic.
            unsafe {
                dst_words
                    .add(i)
                    .write(AtomicU64::from_ptr(src.add(i) as *mut u64).load(Ordering::Relaxed))
            };
        }
        fence(Ordering::Acquire); // end-read: payload loads complete before the re-check
        let s2 = seq.load(Ordering::Relaxed);
        if s1 != s2 {
            continue;
        }
        dst.header_mut().seq = s1;
        dst.header_mut()._pad = 0;
        if dst.header().magic != T::MAGIC {
            return Err(ReadError::MagicMismatch);
        }
        if dst.header().version != T::VERSION {
            return Err(ReadError::VersionMismatch);
        }
        return Ok(());
    }
    Err(ReadError::Busy)
}

File: csharp/PlcBridge/Shm/Seqlock.cs
using System.Runtime.CompilerServices;

namespace PlcBridge.Shm;

public enum SeqlockResult
{
    Ok,
    Busy,            // seqlock: writer kept us spinning
    MagicMismatch,   // wrong segment or layout
    VersionMismatch, // rebuild PLC and bridge with matching layout
}

/// <summary>
/// The seqlock protocol over a shared-memory segment — the C# counterpart of
/// backend/internal/shm/reader.go and writer.go. The protocol (odd = write in
/// progress, even = stable, retry on seq change) and its documented
/// memory-ordering constraints are identical across the Go / Rust / C#
/// implementations; only the atomic primitives differ.
/// </summary>
public static class Seqlock
{
    internal const int MaxRetries = 100;

    /// <summary>Log-parity error strings (reader.go's sentinel errors).</summary>
    public static string Message(SeqlockResult r) => r switch
    {
        SeqlockResult.Busy => "seqlock: writer kept us spinning",
        SeqlockResult.MagicMismatch => "magic mismatch: wrong segment or layout",
        SeqlockResult.VersionMismatch => "version mismatch: rebuild PLC and bridge with matching layout",
        _ => "ok",
    };

    /// <summary>
    /// Copies the segment into dst under the seqlock protocol. On Ok,
    /// dst.Header.Magic and dst.Header.Version are guaranteed to match this
    /// build's expectations.
    ///
    /// Memory-ordering caveat: only Header.Seq is accessed atomically
    /// (Volatile.Read = acquire per ECMA-335); the payload is a plain bulk
    /// copy. Acquire ordering stops the payload's loads from moving above the
    /// first seq load, but nothing stops the hardware from reordering them
    /// past the second seq load — that requires a read fence this code does
    /// not issue. On x86/amd64 (TSO: loads are not reordered with loads) this
    /// is sound, and amd64 is the only deploy target — same constraint as the
    /// Go implementation (see reader.go). Porting to ARM needs
    /// Interlocked.MemoryBarrier() before the second seq load or per-word
    /// atomic copies. The Seq accesses themselves must stay Volatile: they
    /// also prevent the JIT from reordering across them.
    /// </summary>
    public static unsafe SeqlockResult ReadPlcData(Mapping m, out PlcData dst)
    {
        dst = default;
        var src = (PlcData*)m.Ptr;
        ref uint seq = ref Unsafe.AsRef<uint>(&src->Header.Seq);
        for (int i = 0; i < MaxRetries; i++)
        {
            uint s1 = Volatile.Read(ref seq);
            if ((s1 & 1) != 0)
                continue;
            dst = *src;
            uint s2 = Volatile.Read(ref seq);
            if (s1 != s2)
                continue;
            if (dst.Header.Magic != Layout.PlcDataMagic)
                return SeqlockResult.MagicMismatch;
            if (dst.Header.Version != Layout.PlcDataVersion)
                return SeqlockResult.VersionMismatch;
            return SeqlockResult.Ok;
        }
        return SeqlockResult.Busy;
    }

    /// <summary>
    /// Publishes src to the segment under the seqlock protocol. Caller must
    /// initialise src.Header.Magic/Version before the first call (Sink's
    /// constructor does this). Adding fields to PlcCommand requires no changes
    /// here.
    ///
    /// The writer — not the caller — owns Header.Seq. The bulk copy below
    /// would otherwise overwrite the segment's seq with src.Header.Seq (an
    /// even value), briefly making the segment look "stable" while the payload
    /// is still being written, so a reader could latch a torn snapshot. We
    /// force the copied seq to the in-progress (odd) value to keep the seqlock
    /// invariant: seq stays odd for the entire time the payload is in flux.
    ///
    /// Memory-ordering caveat: the payload copy is plain stores between the
    /// two Volatile.Write (release) seq stores. Release ordering stops the
    /// payload's stores from moving below the final seq store, but not from
    /// moving above the first one. Sound on x86/amd64 (TSO: stores are not
    /// reordered with stores), which is the only deploy target — see the
    /// matching note on ReadPlcData before porting to ARM.
    /// </summary>
    public static unsafe void WritePlcCommand(Mapping m, in PlcCommand src)
    {
        var dst = (PlcCommand*)m.Ptr;
        ref uint seq = ref Unsafe.AsRef<uint>(&dst->Header.Seq);
        uint s = Volatile.Read(ref seq);
        // Crash recovery: if the previous writer process died mid-write, the
        // segment's seq is still odd. A plain s+1 would then be even — the
        // "write in progress" marker would look stable and a reader could
        // latch a torn snapshot. Skip ahead to the next odd value instead.
        uint odd = s + 1 + (s & 1);
        Volatile.Write(ref seq, odd); // odd: write in progress

        PlcCommand tmp = src;
        tmp.Header.Seq = odd; // copy carries the in-progress seq, never an even one
        *dst = tmp;

        Volatile.Write(ref seq, odd + 1); // even: stable
    }
}
